lib/src/taiko/host.rs

```diff
@@ -197,37 +197,115 @@ fn execute_data<N: NetworkStrategyBundle<TxEssence = EthereumTxEssence>>(
 
 const BLOB_FIELD_ELEMENT_NUM: usize = 4096;
 const BLOB_FIELD_ELEMENT_BYTES: usize = 32;
-const BLOB_DATA_LEN: usize = BLOB_FIELD_ELEMENT_NUM * BLOB_FIELD_ELEMENT_BYTES;
-
-fn decode_blob_data(blob: &str) -> Vec<u8> {
-    let origin_blob = hex::decode(blob.to_lowercase().trim_start_matches("0x")).unwrap();
-    let header: U256 = U256::from_big_endian(&origin_blob[0..BLOB_FIELD_ELEMENT_BYTES]); // first element is the length
-    let expected_len = header.as_usize();
-
-    assert!(origin_blob.len() == BLOB_DATA_LEN);
-    // the first 32 bytes is the length of the blob
-    // every first 1 byte is reserved.
-    assert!(expected_len <= (BLOB_FIELD_ELEMENT_NUM - 1) * (BLOB_FIELD_ELEMENT_BYTES - 1));
-    let mut chunk: Vec<Vec<u8>> = Vec::new();
-    let mut decoded_len = 0;
-    let mut i = 1;
-    while decoded_len < expected_len && i < BLOB_FIELD_ELEMENT_NUM {
-        let segment_len = if expected_len - decoded_len >= 31 {
-            31
-        } else {
-            expected_len - decoded_len
-        };
-        let segment = &origin_blob
-            [i * BLOB_FIELD_ELEMENT_BYTES + 1..i * BLOB_FIELD_ELEMENT_BYTES + 1 + segment_len];
-        i += 1;
-        decoded_len += segment_len;
-        chunk.push(segment.to_vec());
+const BLOB_DATA_CAPACITY: usize = BLOB_FIELD_ELEMENT_NUM * BLOB_FIELD_ELEMENT_BYTES;
+const BLOB_VERSION_OFFSET: usize = 1;
+const BLOB_ENCODING_VERSION: u8 = 0;
+const MAX_BLOB_DATA_SIZE: usize = (4 * 31 + 3) * 1024 - 4;
+
+// decoding https://github.com/ethereum-optimism/optimism/blob/develop/op-service/eth/blob.go
+fn decode_blob_data(blob_str: &str) -> Vec<u8> {
+    let blob_buf: Vec<u8> = match hex::decode(blob_str.to_lowercase().trim_start_matches("0x")) {
+        Ok(b) => b,
+        Err(_) => return Vec::new(),
+    };
+
+    // check the version
+    if blob_buf[BLOB_VERSION_OFFSET] != BLOB_ENCODING_VERSION {
+        return Vec::new();
+    }
+
+    // decode the 3-byte big-endian length value into a 4-byte integer
+    let output_len =
+        ((blob_buf[2] as u32) << 16 | (blob_buf[3] as u32) << 8 | (blob_buf[4] as u32)) as usize;
+    if output_len > MAX_BLOB_DATA_SIZE {
+        return Vec::new();
+    }
```

Comment:
So version is at byte 1 and length is only 3 bytes? If so, slightly different from the figure in the notion I think. Though avoiding byte 0 makes sense, even though you could argue the version could be stored in just 6 bits, but doesn't matter, just trying to understand and used your figure to do that!

Reply:
oh yes, by design the version & len byte is buf[1..5], not [0..5], I will update the pic.